src/wp-includes/class-wp-scripts.php

```diff
@@ -1058,119 +899,106 @@ private function is_delayed_strategy( $strategy ) {
-				if ( $this->has_inline_script( $dependent ) ) {
-					return false;
-				}
-			} else {
-				// If the dependent script is not using the defer or async strategy, no script in the chain is delayed.
-				$strategy = $this->get_data( $dependent, 'strategy' );
-				if ( $async_only ?
-					'async' !== $strategy :
-					! $this->is_delayed_strategy( $strategy )
-				) {
-					return false;
-				}
-			}
+	private function get_eligible_loading_strategy( $handle ) {
+		$eligible = $this->filter_eligible_strategies( $handle );
 
-			// Recursively check all dependents.
-			if ( ! $this->has_only_delayed_dependents( $dependent, $async_only, $checked ) ) {
-				return false;
-			}
+		// Bail early once we know the eligible strategy is blocking.
+		if ( empty( $eligible ) ) {
+			return '';
 		}
 
-		return true;
+		return in_array( 'async', $eligible, true ) ? 'async' : 'defer' ;
 	}
 
 	/**
-	 * Gets the best eligible loading strategy for a script.
+	 * Filter the list of eligible loading strategies for a script.
 	 *
 	 * @since 6.3.0
 	 *
-	 * @param string  $handle The script handle.
-	 * @return string $strategy The best eligible loading strategy.
+	 * @param string              $handle   The script handle.
+	 * @param string[]|null       $eligible Optional. The list of strategies to filter. Default null.
+	 * @param array<string, true> $checked  Optional. An array of already checked script handles, used to avoid recursive loops.
+	 * @return string[] A list of eligible loading strategies that could be used.
 	 */
-	private function get_eligible_loading_strategy( $handle ) {
+	private function filter_eligible_strategies( $handle, $eligible = null, $checked = array() ) {
+		// If no strategies are being passed, all strategies are eligible.
+		if ( null === $eligible ) {
+			$eligible = $this->delayed_strategies;
+		}
+
+		// If this handle was already checked, return early.
+		if ( isset( $checked[ $handle ] ) ) {
+			return $eligible;
+		}
+
+		// Mark this handle as checked.
+		$checked[ $handle ] = true;
+
+		// If this handle isn't registered, don't filter anything and return.
 		if ( ! isset( $this->registered[ $handle ] ) ) {
-			return '';
+			return $eligible;
+		}
+
+		// If the handle is not enqueued, don't filter anything and return.
+		if ( ! $this->query( $handle, 'enqueued' ) ) {
+			return $eligible;
 		}
 
+		$is_alias = (bool) ! $this->registered[ $handle ]->src;
 		$intended_strategy = $this->get_data( $handle, 'strategy' );
 
-		/*
-		 * Handle known blocking strategy scenarios.
-		 * - An empty strategy is synonymous with blocking.
-		 * - A script alias (where $src is false) must always be blocking since the after inline script cannot be
-		 *   delayed as there is no external script tag and thus no load event at which the inline script can be run.
-		 */
-		if ( empty( $intended_strategy ) ) {
-			return '';
+		// For non-alias handles, an empty intended strategy filters all strategies.
```

Comment:
What does that mean? I'm not sure I understand. Does returning an empty array here mean that the only eligible strategy is "blocking"? Or does it mean any strategy is eligible? If the latter, I think it would be more intuitive here to just return the array that includes `defer` and `async`, or not?

Reply:
Right. An empty array means that the script should be blocking. However, an alias (I.e., a handle with no `src`) will not have a strategy applied, so we skip over the check in that case.

Reply:
Thanks @joemcgill, makes more sense now!